### ISSUE TITLE
A simple benchmark for uftrace record

### ISSUE THREAD

**Issue by namhyung** [namhyung/uftrace#1602]
Hello,

I'm preparing a simple benchmark program and script to measure the overhead of uftrace record.  The program is really simple that calls very small functions in a loop.

```c
int foo(volatile int *ptr)
{
	*ptr += 1;
	return *ptr;
}

int baz(volatile int *ptr)
{
	*ptr += 1;
}

int bar(volatile int *ptr)
{
	baz(ptr);
	return *ptr;
}

int bench(int count)
{
	volatile int result = 0;

	for (int i = 0; i < count; i++) {
		if (i % 2 == 0)
			foo(&result);
		else
			bar(&result);
	}
	return result;
}
```
Then uftrace record and report to show the avg and minimum self-time.  We can use it when we'd like to make some changes that will affect the performance like add some filters or triggers.

The `misc/bench.sh` script is to run this.

```
$ misc/bench.sh -h
Usage: bench.sh [OPTION]
  OPTION  -c N      Use CPU N during the benchmark. (default: 3)
          -p PROG   Use program PROG. (default: ./bench)
          -u UOPT   Use uftrace option UOPT. Please quote it.
          -v        Show verbose messages.
          -h        Show this help and exit.
```

Running it simply will show the result like this (please do it in the `misc` directory).  It needs `sudo` because it would like to set the cpufreq governor for deterministic performance.
```
$ cd misc
$ sudo ./bench.sh
# uftrace bench
    Self avg    Self min  Function
  ==========  ==========  ====================
    0.112 us    0.071 us  bar
    0.029 us    0.026 us  foo
    0.029 us    0.025 us  baz
```

Let's run it with a different option
```
$ sudo ./bench.sh -u '-F main'
# uftrace bench
    Self avg    Self min  Function
  ==========  ==========  ====================
    0.133 us    0.087 us  bar
    0.033 us    0.029 us  baz
    0.033 us    0.029 us  foo
```

**Comment by namhyung** [comment]
Here's the pseudo code for the benchmark functions:

       foo():                        <libmcount.so>
    A    // prologue
    B    // calls mcount
                                    mcount():
    C                                  // save registers/arguments
    D                                  // check filters/triggers
    E                                  // setup rstack
    F                                  // save timestamp       <--------------+
    G                                  // hook return address                 |
    H                                  // restore registers/arguments         |
    I                                  // go back to the original function    |
    J    // do the function body                                              | = U_a
    K    // return                                                            |
                                    mcount_exit()                             |
    L                                  // do some minimal checking            |
    M                                  // save timestamp       <--------------+
    N                                  // check filters/triggers
    O                                  // save rstacks to shmem
    P                                  // return to the original caller
    
Note that uftrace can only record time from step F to step M for the
function foo(). For the steps C, D, E and N, O, P, we need to use nested
functions like bar() and baz().

For simplicity, let's call the portion of overhead in uftrace that's
covered by the leaf function (F-I and L-M) as `U_a`, and rest of them as
`U_b` (for C-E and N-P).  Also let's say the overhead of the original
function as O.  So the performance of each functions should be:
    
      foo   = O_foo + U_a + U_b
                            ^^^ (not measured)
      ${foo_self} = O_foo + U_a
    
We want to measure the performance of 'foo' but what uftrace shows is
just actually `O_foo + U_a` and we don't know U_b.  We know O_foo is
very small as it's a leaf function with a memory operation for the same
address (it should hit the L1 cache most likely).
    
To determine the U_b part, we can use the nested functions:
    
      bar   = O_bar + U_a + U_b + ${baz}
        baz = O_baz + U_a + U_b

But in this case, what uftrace shows for bar(total) is actually:
    
      ${bar_total} = O_bar + U_a + (O_baz + U_a + U_b)
                                    -----------
                                    ${baz_self}
      O_bar ~= O_baz
    
      ${bar_self} = ${bar_total} - ${baz_total}
                  = ${bar_total} - ${baz_self}  // baz() is a leaf
                  = O_bar + U_a + U_b
                 ~= O_baz + U_a + U_b
                  = ${baz_self} + U_b
    
      U_b = ${bar_self} - ${baz_self}
    
Now we can get the `U_b` part, but it still has some issue.  Because the
step O (save rstacks to shmem) contains some records for the parent
functions as uftrace postpones the recording of entry functions to the
exit functions.  So step O in the baz function will record 3 data:
    
      1 entry of bar + 1 entry of baz + 1 exit of baz

But we cannot know exact overhead of the step O so we will ignore the
issue for now.  So the self-time of bar() already contains `U_b` part
and eliminated self-time of baz().  Then it's the precise overhead of
a single function for uftrace.  It's around 70 nsec at minimum on my laptop.

**Comment by honggyukim** [comment]
Hi @namhyung,

I couldn't look into the detail, but at the first glance, my linux environment running on parallels under M1 macbook shows the following error. We need a logic to skip if `/sys/devices/system/cpu/cpufreq/policy3/scaling_governor` isn't available.
```
$ make bench
  CC       uftrace.o
  LINK     uftrace

# uftrace bench
cat: /sys/devices/system/cpu/cpufreq/policy3/scaling_governor: No such file or directory
taskset: failed to set pid 55429's affinity: Invalid argument
WARN: cannot open record data: bench.data: No such file or directory
```
I've commented out lines for setting cpufreq, then see the output as follows.
```
$ make bench
  CC       uftrace.o
  LINK     uftrace

# uftrace bench
    Self avg    Self min  Function
  ==========  ==========  ====================
    0.099 us    0.041 us  bar
    0.043 us    0.041 us  foo
    0.043 us    0.041 us  baz
   50.540 us    2.500 us  linux:schedule (pre-empted)
```
However, I see some weird result, which shows empty self field as follows.
```
$ make bench
  CC       uftrace.o
  LINK     uftrace

# uftrace bench
    Self avg    Self min  Function
  ==========  ==========  ====================
    0.100 us              bar
    0.043 us    0.041 us  foo
    0.043 us    0.041 us  baz
   39.166 us   39.166 us  linux:schedule (pre-empted)
```
This problem was observed before, but need more investigation.

**Comment by namhyung** [comment]
Ok, it seems like a clock resolution problem.  I guess it's because you run it on a virtual machine.  Maybe we need to run it on a physical hw.

**Comment by namhyung** [comment]
Updated [review/bench](../compare/review/bench) branch.

**Comment by ParkHanbum** [comment]
```
m@M:~/uftrace/misc$ uftrace report
  Total time   Self time       Calls  Function
  ==========  ==========  ==========  ====================
  148.939 ms    0.200 us           1  main
  148.939 ms   64.784 ms           1  bench
   64.289 ms   52.891 ms      480278  bar
   19.865 ms   19.865 ms      173485  foo
   11.376 ms   11.376 ms       99365  baz
   21.100 us   21.100 us           2  linux:schedule (pre-empted)
    0.600 us    0.600 us           1  __monstartup
    0.300 us    0.300 us           1  __cxa_atexit
m@M:~/uftrace/misc$ uftrace report -f self-avg,self-min bench
    Self avg    Self min  Function
  ==========  ==========  ====================
    0.200 us    0.200 us  main
   64.784 ms   64.784 ms  bench
    0.110 us              bar
    0.114 us    0.100 us  foo
    0.114 us    0.100 us  baz
   10.550 us   10.300 us  linux:schedule (pre-empted)
    0.600 us    0.600 us  __monstartup
    0.300 us    0.300 us  __cxa_atexit
m@M:~/uftrace/misc$
```
why I can't see self min of bar?

**Comment by namhyung** [comment]
I guess it's 0.  If the total time of `bar` is 0.1 us, the self time would be 0 because `bar` calls `baz` and `baz` takes 0.1 us.  It seems the clock cannot return values smaller than 0.1 us on your system.

**Comment by namhyung** [comment]
Merged e0f8aa2d33.
